### PR TITLE
feat: stmgr: cache migrated stateroots

### PR DESCRIPTION
**Pull request by travisperson**
## Related Issues
#10048

## Proposed Changes
- Cache migration results to avoid running migrations twice when making calls that execute an upgrade epoch

## Checklist

Before you mark the PR ready for review, please make sure that:

- [ ] Commits have a clear commit message.
- [ ] PR title is in the form of of `<PR type>: <area>: <change being made>`
  - example: ` fix: mempool: Introduce a cache for valid signatures`
  - `PR type`: fix, feat, build, chore, ci, docs, perf, refactor, revert, style, test
  - `area`, e.g. api, chain, state, market, mempool, multisig, networking, paych, proving, sealing, wallet, deps
- [ ] New features have usage guidelines and / or documentation updates in
  - [ ] [Lotus Documentation](https://lotus.filecoin.io)
  - [ ] [Discussion Tutorials](https://github.com/filecoin-project/lotus/discussions/categories/tutorials)
- [ ] Tests exist for new functionality or change in behavior
- [ ] CI is green

### CODE REVIEW

File: chain/stmgr/stmgr.go [hunk 1]
```diff
@@ -51,9 +51,10 @@ type versionSpec struct {
 }
 
 type migration struct {
-	upgrade       MigrationFunc
-	preMigrations []PreMigration
-	cache         *nv16.MemMigrationCache
+	upgrade            MigrationFunc
+	preMigrations      []PreMigration
+	cache              *nv16.MemMigrationCache
+	migratedStateroots map[cid.Cid]cid.Cid
```
Comment (travisperson) [inline chain/stmgr/stmgr.go:57]: Map here is for reorgs that might make us run the same migration on different stateroots, we want to make sure we grab the correct migrated root for each input.

File: chain/stmgr/forks.go [hunk 2]
```diff
@@ -174,9 +174,16 @@ func (us UpgradeSchedule) GetNtwkVersion(e abi.ChainEpoch) (network.Version, err
 
 func (sm *StateManager) HandleStateForks(ctx context.Context, root cid.Cid, height abi.ChainEpoch, cb ExecMonitor, ts *types.TipSet) (cid.Cid, error) {
 	retCid := root
-	var err error
 	u := sm.stateMigrations[height]
 	if u != nil && u.upgrade != nil {
+		migCid, ok, err := u.resultCache.Result(ctx, root)
+		if err == nil && ok {
+			log.Warnw("CACHED migration", "height", height, "from", root, "to", migCid)
```
Comment (Stebalien) [inline chain/stmgr/forks.go:181]: This probably doesn't warrant a warning.

File: chain/stmgr/stmgr.go [hunk 3]
```diff
@@ -54,6 +56,48 @@ type migration struct {
...
 	cache         *nv16.MemMigrationCache
+	resultCache   *resultCache
+}
+
+type resultCache struct {
+	ds        dstore.Batching
+	keyPrefix string
+}
+
+func (m *resultCache) Result(ctx context.Context, root cid.Cid) (cid.Cid, bool, error) {
```
Comment (Stebalien) [inline chain/stmgr/stmgr.go:67]: super-nit: call this `Get`? When I see `Result`, I'm not sure if we're putting the result or getting it.

File: chain/stmgr/stmgr.go [hunk 4]
```diff
@@ -54,6 +56,48 @@ type migration struct {
 	upgrade       MigrationFunc
 	preMigrations []PreMigration
 	cache         *nv16.MemMigrationCache
+	resultCache   *resultCache
+}
+
+type resultCache struct {
```
Comment (Stebalien) [inline chain/stmgr/stmgr.go:62]: super-nit: migrationResultCache, migrationCache, etc. or something like that?

File: chain/stmgr/stmgr.go [hunk 5]
```diff
@@ -54,6 +56,48 @@ type migration struct {
...
+	ds        dstore.Batching
+	keyPrefix string
+}
+
+func (m *resultCache) Result(ctx context.Context, root cid.Cid) (cid.Cid, bool, error) {
+	kStr := fmt.Sprintf("%s-%s", m.keyPrefix, root)
+	k := dstore.NewKey(kStr)
```
[the remark this answers is not in the document]
Comment (Stebalien) [reply chain/stmgr/stmgr.go:69]: This seems reasonable as-is. Although:

- I'm not sure how useful the epoch is.
- We usually use slashes for key separators: `$nv/$root`.
- I'd add a general `/migration-cache/` prefix at the beginning.

File: chain/stmgr/stmgr.go [hunk 6]
```diff
@@ -54,6 +56,48 @@ type migration struct {
...
+	}
+
+	if !found {
+		return cid.Undef, false, nil
+	}
+
+	bs, err := m.ds.Get(ctx, k)
+	if err != nil {
+		return cid.Undef, false, xerrors.Errorf("error loading migration result: %w", err)
+	}
```
Comment (Stebalien) [inline chain/stmgr/stmgr.go:83]: Do this in one go and check `ipld.IsNotFound(err)`.

File: chain/stmgr/stmgr.go [hunk 7]
```diff
@@ -54,6 +56,48 @@ type migration struct {
...
+	if err != nil {
+		return cid.Undef, false, xerrors.Errorf("error parsing migration result: %w", err)
+	}
+
+	return c, true, nil
+}
+
+func (m *resultCache) Store(ctx context.Context, root cid.Cid, resultCid cid.Cid) error {
+	kStr := fmt.Sprintf("%s-%s", m.keyPrefix, root)
+	k := dstore.NewKey(kStr)
```
Comment (Stebalien) [inline chain/stmgr/stmgr.go:95]: nit: helper? `keyForMigration`?

File: chain/stmgr/forks.go (hunk 2 above)
Comment (travisperson) [reply chain/stmgr/forks.go:181]: Ya, it's only a warning because everything else is a warning eg 'STARTING migration', 'COMPLETED migraiton'.

Comment (Stebalien) [reply chain/stmgr/forks.go:181]: Warning the first time we migrate makes sense, but warning when we fetch a cached migration doesn't.

File: chain/stmgr/stmgr.go [hunk 8]
```diff
@@ -51,9 +54,44 @@ type versionSpec struct {
...
+	c, err := cid.Parse(bs)
+	if err != nil {
+		return cid.Undef, false, xerrors.Errorf("error parsing migration result: %w", err)
+	}
+
+	return c, true, nil
+}
+
+func (m *migrationResultCache) Store(ctx context.Context, root cid.Cid, resultCid cid.Cid) error {
+	kStr := fmt.Sprintf("%s-%s", m.keyPrefix, root)
```
Comment (Stebalien) [inline chain/stmgr/stmgr.go:88]: The key format differs. I'd make that helper function.

**Review by Stebalien — APPROVED** [review]

**Review by arajasek — COMMENTED** [review]
LGTM!

**Pull request merged**